### DISCUSSION TITLE
Akvorado on kubernetes, sFlow exporter on linux returning empty ifDescr and no flows are getting registered

### DISCUSSION THREAD

**Discussion by nkichukov** [akvorado/akvorado#1442] [Q&A] [answered]
Hi folks,
This is about the latest version of akvorado setup on kubernetes cluster, where the exporters are linux servers with multiple interfaces that run net-snmpd daemon. The sFlow exporter is pmacctd, took a while to configure it so that the flows appeared in the inlet under /api/v0/inlet/flows endpoint, but now they are flowing, example:
`{"TimeReceived":1728670170,"SamplingRate":100,"ExporterAddress":"::ffff:192.168.1.10","InIf":0,"OutIf":4,"SrcVlan":0,"DstVlan":0,"SrcAddr":"::ffff:<redactedPublicIP>","DstAddr":"::ffff:<redactedPublicIP>","NextHop":"","SrcAS":0,"DstAS":0,"GotASPath":false,"SrcNetMask":0,"DstNetMask":0}`

The issue is that the flows arrive in the `inlet` but there is nothing stored in the other layers: kafka and clickhouse, so the UI shows 0 flows.

One theory is that the SNMP poller is not able to receive the interface description, which is non-existent on this snmpd. I am trying to configure snmpd to return something for ifDescr on these oids for the interfaces, but have not had any success yet.

As far as I understood, if the description is not set, the flows are not registered.

I've read on the discussions that it is possible to configure the exporters so that the interface index with the needed description can be hardcoded in the inlet, but I am struggling to find exactly where, in which yaml config is this supposed to be set? There is a reference that it exists in the documentation, but I cannot find such example.

the error metrics look like this, coming from 3 exporters:
```
/api/v0/inlet/metrics | grep '^akvorado_inlet' | grep _error 

akvorado_inlet_core_flows_errors_total{error="input and output interfaces missing",exporter="192.168.1.10"} 2132
akvorado_inlet_core_flows_errors_total{error="input and output interfaces missing",exporter="192.168.1.11"} 318
akvorado_inlet_core_flows_errors_total{error="input and output interfaces missing",exporter="192.168.1.12"} 200
```
Can I get an example snippet and directions on where to hardcode exporter setup to be used by the inlet?

Thank you!

**Comment by vincentbernat** [comment, marked as answer]
There is an example in the [documentation](https://demo.akvorado.net/docs/configuration#metadata). Search for "static provider". You need to put that in `inlet.yaml`.

**Reply by nkichukov** [reply]
Hello Vincent,
Thank you for the directions. Indeed, the static provider seems to be clearly defined, wonder why I missed it before?
Anyways, the 3 exporters continue to be generating the error in the `inlet` as `input and output interfaces missing` as if the static exporters configuration under metadata made no difference?

I've tried `use-src-addr-for-exporter-addr: true` on the sflow listener to see if there is an issue with overriding the original IP with an `agent_ip` on pmacctd, but this did not change the result.

I've set the static provider to be on top of the list, before `snmp` to make it be tried first, however no luck.

In the flows, I can see the exporter IP is prefixed with ::ffff:, tried that. Tried with /32, /24, without network mask in the CIDR. Tried with fixed exporter IP(not the agent IP which differ, the agent IP is 192.168.1.0/24, which is accessible to the `inlet` on the reverse path), tried with the agent IP, absolutely no difference. Tried with 0.0.0.0/0 - no change. Tried defining all 3 exporters, no luck either.

What else could be going wrong here? The flows keep arriving, but are not imported in akvorado.

Are these exporters/agent IPs needed to be registered elsewhere or what else can I be missing?

Here is the complete metadata configuration with all 3 exporters configured under `static` provider:

<details>
<code>
    metadata:
      workers: 10
      providers:
        - type: static
          exporters:
            192.168.1.10:
              name: rtr-0
              default:
                name: unknown
                description: Unknown interface
                speed: 1000
              ifindexes:
                0:
                  name: lo
                  description: Loopback
                  speed: 100000
                1:
                  name: eth0
                  description: Kubernetes pod network
                  speed: 100000
                4:
                  name: eifa
                  description: External interface a
                  speed: 25000
                5:
                  name: eifb
                  description: External interface b
                  speed: 25000
                6:
                  name: if-v4
                  description: Internet IPv4
                  speed: 25000
                7:
                  name: if-v6
                  description: Internet IPv6
                  speed: 25000
                8:
                  name: uplink
                  description: uplink
                  speed: 25000
            192.168.1.11:
              name: rtr-1
              default:
                name: unknown
                description: Unknown interface
                speed: 1000
              ifindexes:
                0:
                  name: lo
                  description: Loopback
                  speed: 100000
                1:
                  name: eth0
                  description: Kubernetes pod network
                  speed: 100000
                4:
                  name: eifa
                  description: External interface a
                  speed: 25000
                5:
                  name: eifb
                  description: External interface b
                  speed: 25000
                6:
                  name: if-v4
                  description: Internet IPv4
                  speed: 25000
                7:
                  name: if-v6
                  description: Internet IPv6
                  speed: 25000
                8:
                  name: uplink
                  description: uplink
                  speed: 25000
            192.168.1.12:
              name: rtr-2
              default:
                name: unknown
                description: Unknown interface
                speed: 1000
              ifindexes:
                0:
                  name: lo
                  description: Loopback
                  speed: 100000
                1:
                  name: eth0
                  description: Kubernetes pod network
                  speed: 100000
                4:
                  name: eifa
                  description: External interface a
                  speed: 25000
                5:
                  name: eifb
                  description: External interface b
                  speed: 25000
                6:
                  name: if-v4
                  description: Internet IPv4
                  speed: 25000
                7:
                  name: if-v6
                  description: Internet IPv6
                  speed: 25000
                8:
                  name: uplink
                  description: uplink
                  speed: 25000
        - type: snmp
          communities:
            ::/0: redacted
</code>
</details>

With every change, I've restarted the orchestrator and inlet service in sequence, still no luck.

Anything else to try?

**Reply by vincentbernat** [reply]
Your configuration looks correct. You are sure the file is used? For example, if you introduce a syntax error?

**Reply by nkichukov** [reply]
It is being used, if I introduce a syntax error, first the orchestrator fails to start.
Then when I reconfigure the listeners, for example `use-src-addr-for-exporter-addr` it is reflected immediately upon restart. Same with the community string for the snmp provider, which I can see change when dumping traffic (when not using static). 

Can debug logs be enabled on the `inlet` so it can dump more information on what's going on with this metadata provider?

Right now it only prints the :8080 `curl` accesses dumping metrics/flows.

**Reply by vincentbernat** [reply]
Are the errors about input and output interfaces missing still increasing? The `/api/v0/inlet/flows` endpoint exports flows just before Kafka, so if they look correct, the problem could be elsewhere.

**Reply by nkichukov** [reply]
Yes, they continue to increase. The number of UDP packets total is not equal to the number of flow errors with missing input and output interfaces names. If the flows arrive and inlet is doing its job, what is the next component to handle the flows before they show in the console?

**Reply by nkichukov** [reply]
PVC for clickhouse storage under /var/lib/clickhouse got full. I will increase this and restart to see if this helps.

**Reply by vincentbernat** [reply]
As long as errors are increasing, it means some flows are missing. As you have a default for the static provider, I suppose this is because some routers are not defined in your configuration? You can also use ::/0 as a catch all (both v4 and v6).

**Reply by nkichukov** [reply]
I've added the `catch all` static provider address on top of all others, with `::/0` and name of `rtr-catchall` with default entry and `ifindexes`, same as the other entries. The udp packets arrive, but the same errors for `akvorado_inlet_core_flows_errors_total` keep showing up and nothing is registered in the UI.

I've even cleaned up kafka and zookeeper topics and all the data they've stored to start anew, I've let the orchestrator recreate the topic on restart.

Still nothing.

I wonder if I should try going back to netflow on the exporters (I've switched to sFlow because it supported feeding `agent_ip` which could then be used for the SNMP provider to fetch data, but now with the static provider it is no longer needed)?

Would netflow make any difference versus sFlow in this case?

Thank you,
-Nikolay

**Comment by vincentbernat** [comment]
You added "0" as ifIndex. Is that something you really have in your sFlow packets? An interface index should be between 1 and 2147483647. No lookup is done for index 0 as it is considered as missing.

**Reply by nkichukov** [reply]
Hello Vincent,

Your direction towards looking elsewhere (after inlet) got me review what appears to have been problematic kafka/zookeeper setup. I've cleaned up the flow topics, rebuilt kafka and zookeeper and let orchestrator create a new topic. Then once kafka and zookeeper were happy I could see that some packets from the `flow` started showing up in `clickhouse` and thus the `console`.

So the static providers worked, which is great, as I could remove the `snmpd` service on the exporters completely(I deployed it only for akvorado, for the missing description).

I've kept the `catch-all` static provider, just in case, though in a new deployment (another datacenter), I only get to see 3 registered exporters, so there is no packet going into the `catch-all` provider.

Now, your observation with ifIndex 0 is correct, I saw that `inlet` was complaining about this, so I've set the configuration start at `1`.

However, the errors with missing interfaces/description continue to increase... so there must be something else on the exporters.

This is what appears in the flow and I believe is classified and thus makes it into clickhouse (taken from earlier posts):
```
{"TimeReceived":1728670170,"SamplingRate":100,"ExporterAddress":"::ffff:192.168.1.10","InIf":0,"OutIf":4,"SrcVlan":0,"DstVlan":0,"SrcAddr":"::ffff:<redactedPublicIP>","DstAddr":"::ffff:<redactedPublicIP>","NextHop":"","SrcAS":0,"DstAS":0,"GotASPath":false,"SrcNetMask":0,"DstNetMask":0}
```

According to this, `InIf` is 0. Does this mean that akvorado is not classifying this packet as it has index 0? Maybe this is why the error counters continue to increase, if the exporters are sending some of the flows with interface index 0?

No matter what, it seems like I am not able to observe any flows being lost(not that I've done a in-depth analysis) but at a first glance, any traffic I generate that flows via the routers is registered in akvorado, though error counters are increasing?

If we go to the exporters, this is `pmacctd` software that makes use of:
```
sfprobe_direction[any]: tag
sfprobe_ifindex[any]: tag2
pre_tag_map: /etc/pretag.map
pcap_interfaces_map: /etc/interfaces.map
```

`tag` is for the direction and `tag2` is for the index. `Tag2` starts at index `1`, so there should be no `InIf:0` in the flows, is this correct assumption?

Also, to make it pickup all the router interfaces, I've defined them in the interfaces.map, which has something similar to:
```
ifindex=100  ifname=lo
!ifindex=300  ifname=eth0
ifindex=400  ifname=eifa
ifindex=500  ifname=eifb
... more interfaces ...
```

Switching those to their SNMP indexes, 1, 4, 5, etc changes nothing for the `akvorado` error counters.

I know `pmacctd` is supposed to be discussed elsewhere, but I would have appreciated it if `akvorado` documentation had a working example for getting started for it as well, such as the various proprietary router configurations that exist today.

As it is working at the moment for me, I am happy to mark this Q&A topic as resolved, but would be happier if it ended up adding to the Documentation with a simple `pmacctd` exporter example, if possible?

**Reply by vincentbernat** [reply]
Either InIf or OutIf needs to be different of 0. That's because locally generated packets or locally processed packets have an ifindex of 0.

You can propose a PR against the documentation with a pmacct configuration (like this is done for hardware vendors).

**Reply by nkichukov** [reply]
https://github.com/akvorado/akvorado/pull/1455 is now open.
Thanks for all your help, Vincent.
